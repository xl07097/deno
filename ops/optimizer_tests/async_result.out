#[allow(non_camel_case_types)]
///Auto-generated by `deno_ops`, i.e: `#[op]`
///
///Use `op_async_result::decl()` to get an op-declaration
///you can include in a `deno_core::Extension`.
pub struct op_async_result;
#[doc(hidden)]
impl op_async_result {
    pub const fn name() -> &'static str {
        stringify!(op_async_result)
    }
    #[allow(clippy::not_unsafe_ptr_arg_deref)]
    pub extern "C" fn v8_fn_ptr<'scope>(
        info: *const deno_core::v8::FunctionCallbackInfo,
    ) {
        let info = unsafe { &*info };
        let scope = &mut unsafe { deno_core::v8::CallbackScope::new(info) };
        let args = deno_core::v8::FunctionCallbackArguments::from_function_callback_info(
            info,
        );
        let rv = deno_core::v8::ReturnValue::from_function_callback_info(info);
        Self::v8_func(scope, args, rv);
    }
    pub const fn decl<'scope>() -> deno_core::OpDecl {
        deno_core::OpDecl {
            name: Self::name(),
            v8_fn_ptr: Self::v8_fn_ptr as _,
            enabled: true,
            fast_fn: {
                use deno_core::v8::fast_api::Type::*;
                use deno_core::v8::fast_api::CType;
                Some(
                    deno_core::v8::fast_api::FastFunction::new(
                        &[V8Value, Int32, Uint32, CallbackOptions],
                        CType::Void,
                        op_async_result_fast_fn as *const ::std::ffi::c_void,
                    ),
                )
            },
            is_async: true,
            is_unstable: false,
            is_v8: false,
            force_registration: false,
            arg_count: 2usize as u8,
        }
    }
    #[inline]
    #[allow(clippy::too_many_arguments)]
    async fn call(state: Rc<RefCell<OpState>>, rid: ResourceId) -> Result<u32, Error> {}
    pub fn v8_func<'scope>(
        scope: &mut deno_core::v8::HandleScope<'scope>,
        args: deno_core::v8::FunctionCallbackArguments,
        mut rv: deno_core::v8::ReturnValue,
    ) {
        use deno_core::futures::FutureExt;
        let ctx = unsafe {
            &*(deno_core::v8::Local::<deno_core::v8::External>::cast(args.data()).value()
                as *const deno_core::_ops::OpCtx)
        };
        let promise_id = args.get(0);
        let promise_id = deno_core::v8::Local::<
            deno_core::v8::Integer,
        >::try_from(promise_id)
            .map(|l| l.value() as deno_core::PromiseId)
            .map_err(deno_core::anyhow::Error::from);
        let promise_id: deno_core::PromiseId = match promise_id {
            Ok(promise_id) => promise_id,
            Err(err) => {
                deno_core::_ops::throw_type_error(
                    scope,
                    format!("invalid promise id: {}", err),
                );
                return;
            }
        };
        let arg_0 = args.get(1usize as i32);
        let arg_0 = match deno_core::serde_v8::from_v8(scope, arg_0) {
            Ok(v) => v,
            Err(err) => {
                let msg = format!(
                    "Error parsing args at position {}: {}", 1usize,
                    deno_core::anyhow::Error::from(err)
                );
                return deno_core::_ops::throw_type_error(scope, msg);
            }
        };
        let fut = deno_core::_ops::map_async_op1(
            ctx,
            Self::call(ctx.state.clone(), arg_0),
        );
        let maybe_response = deno_core::_ops::queue_async_op(
            ctx,
            scope,
            false,
            promise_id,
            fut,
        );
        if let Some(response) = maybe_response {
            rv.set(response);
        }
    }
}
#[allow(clippy::too_many_arguments)]
fn op_async_result_fast_fn<'scope>(
    _: deno_core::v8::Local<deno_core::v8::Object>,
    __promise_id: i32,
    rid: ResourceId,
    fast_api_callback_options: *mut deno_core::v8::fast_api::FastApiCallbackOptions,
) -> () {
    use deno_core::v8;
    use deno_core::_ops;
    let __opts: &mut v8::fast_api::FastApiCallbackOptions = unsafe {
        &mut *fast_api_callback_options
    };
    let __ctx = unsafe {
        &*(v8::Local::<v8::External>::cast(unsafe { __opts.data.data }).value()
            as *const _ops::OpCtx)
    };
    let state = __ctx.state.clone();
    let result = op_async_result::call(state, rid);
    let result = _ops::queue_fast_async_op(__ctx, __promise_id, result);
}
