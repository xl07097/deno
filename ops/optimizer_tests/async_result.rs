async fn op_async_result(
  state: Rc<RefCell<OpState>>,
  rid: ResourceId,
) -> Result<u32, Error> {
  // @test-attr:fast
}
