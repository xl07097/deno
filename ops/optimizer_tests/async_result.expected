=== Optimizer Dump ===
returns_result: true
has_ref_opstate: false
has_rc_opstate: true
has_fast_callback_option: false
needs_fast_callback_option: false
fast_result: None
fast_parameters: [V8Value, I32, U32]
transforms: {}
is_async: true
fast_compatible: true
